src/main/java/at/hannibal2/skyhanni/features/event/diana/ArrowGuessBurrow.kt

```diff
@@ -0,0 +1,224 @@
+import at.hannibal2.skyhanni.utils.BlockUtils.getBlockAt
+import at.hannibal2.skyhanni.utils.BlockUtils.isInLoadedChunk
+import at.hannibal2.skyhanni.utils.LocationUtils.isInside
+import at.hannibal2.skyhanni.utils.LorenzVec
+import at.hannibal2.skyhanni.utils.RaycastUtils
+import at.hannibal2.skyhanni.utils.collection.TimeLimitedSet
+import net.minecraft.init.Blocks
+import net.minecraft.util.EnumParticleTypes
+import kotlin.time.Duration.Companion.minutes
+
+@SkyHanniModule
+object ArrowGuessBurrow {
+    private val config get() = SkyHanniMod.feature.event.diana
+
+    private var points: HashSet<LorenzVec> = HashSet()
+    private var recentArrowParticles = TimeLimitedSet<LorenzVec>(1.minutes)
+
+    @HandleEvent(onlyOnIsland = IslandType.HUB, receiveCancelled = true)
+    fun onReceiveParticle(event: ReceiveParticleEvent) {
+        if (!isEnabled()) return
+
+        if (event.distanceToPlayer > 6) return
+        if (event.type != EnumParticleTypes.REDSTONE) return
+        if (event.count != 0) return
+        if (event.speed != 1.0f) return
+
+        val allowedOffsets = setOf(0.0, 128.0, 255.0)
+        if (event.offset.x !in allowedOffsets) return
+        if (event.offset.y !in allowedOffsets) return
+        if (event.offset.z !in allowedOffsets) return
+
+        if (recentArrowParticles.contains(event.location)) return
+        recentArrowParticles.add(event.location)
+        points.add(event.location)
+
+        val arrow = detectArrow(points) ?: return
+        val guess = findClosestValidBlockToRay(arrow) ?: return
+
+        // if you dig a burrow while its tracking particles it doesn't create a new waypoint I think but this is rare and non-fatal
+        GriffinBurrowHelper.newBurrow = false
+        BurrowGuessEvent(
+            guess,
+            precise = true,
+            new = true
+        ).post()
+
+        points.clear()
+    }
+
+    private fun findClosestValidBlockToRay(ray: RaycastUtils.Ray): LorenzVec? {
+        val bounds = IslandType.HUB.islandData?.boundingBox
+        val step = 0.9
+
+        if (bounds == null) return null
+
+        var closest: LorenzVec? = null
+        var closestDistance = Double.MAX_VALUE
+
+        val seen = HashSet<LorenzVec>()
+        seen.add(ray.origin.roundToBlock())
+
+        // travel the ray
+        var iteration = 0
+        while (true) {
+            val current = ray.origin + ray.direction.times(step * iteration)
+            if (!bounds.isInside(current)) break
+
+            val pos = current.roundToBlock()
+            val center = pos.add(0.5,0.5,0.5)
+
+            if (seen.add(pos) && isBlockValid(pos)) {
+                val distance = RaycastUtils.findDistanceToRay(ray, center)
+                if (closest == null || closestDistance > distance) {
+                    closest = pos
+                    closestDistance = distance
+                }
+            }
+            iteration++
+        }
```

Comment:
you should add some sort of safety check so the game doesnt deadlock inside the while loop